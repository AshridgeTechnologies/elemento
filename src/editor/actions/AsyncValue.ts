export default class AsyncValue<T> {
    private _value: T | null = null
    private notifyFn?: (value: T) => void
    private promise?: Promise<void>

    get value() { return this._value }

    init(getValueFn: () => Promise<T>, notifyFn: (value: T) => void) {
        if (!this.promise) {
            const valuePromise = getValueFn()
            console.log('valuePromise', valuePromise)
            this.promise = valuePromise.then(value => {
                this._value = value
                this.notifyFn?.(value)
            })
        }
        this.notifyFn = notifyFn
    }
}