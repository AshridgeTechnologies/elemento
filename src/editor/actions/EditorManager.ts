import GitProjectStore from '../GitProjectStore'
import http from 'isomorphic-git/http/web'
import {DiskProjectStore} from '../DiskProjectStore'

import {editorEmptyProject} from '../../model/Project'

export default class EditorManager {
    constructor(private openProjectFunction: (name: string, projectStore: DiskProjectStore) => Promise<void>)
    {}
    getFromGitHub = async (url: string, directory: FileSystemDirectoryHandle) => {
        const projectStore = new DiskProjectStore(directory)
        const gitStore =  new GitProjectStore(projectStore.fileSystem, http)
        await gitStore.clone(url)
        await this.openProjectFunction(directory.name, projectStore)
    }

    newProject = async(directory: FileSystemDirectoryHandle) => {
        const projectStore = new DiskProjectStore(directory)
        await projectStore.createProject()
        await projectStore.writeProjectFile(editorEmptyProject())
        await this.openProjectFunction(directory.name, projectStore)
    }
}