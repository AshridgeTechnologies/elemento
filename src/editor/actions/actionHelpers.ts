import React, {ChangeEvent, useState} from 'react'
import {AlertColor} from '@mui/material'
import EditorManager from './EditorManager'

export type ShowAlertFn = (title: string, message: string, detail: React.ReactNode, severity: AlertColor) => void

export type Optional<T> = {
    [Property in keyof T]+?: T[Property];
}

export function validateProjectName(name: string, existingNames: string[]) {
    if (existingNames.includes(name)) return 'There is already a project with this name'
    const match = name.match(/[^\w \(\)#%&+=.-]/g)
    if (match) return 'Name contains invalid characters: ' + match.join('')
    return null
}

export const onChangeValue = (onChangeFn: (val: string) => void) => (event: ChangeEvent) => onChangeFn((event.target as HTMLInputElement).value)

export class UIManager {
    readonly onClose: () => void
    readonly showAlert: ShowAlertFn
    constructor(props: {onClose: () => void, showAlert: ShowAlertFn}) {
        this.onClose = props.onClose
        this.showAlert = props.showAlert
    }
}

type StateClass<T> = { new(props: { editorManager: EditorManager, uiManager: UIManager }, onUpdate: (t: T) => void): T }

export function useDialogState<T>(stateClass: StateClass<T>, editorManager: EditorManager, uiManager: UIManager) {
    const onUpdate = (newState: T) => setState(newState)
    const [state, setState] = useState(() => new stateClass({editorManager, uiManager}, onUpdate))
    return state
}

export const doAction = (uiManager: UIManager, description: string, actionFn: () => (void)) => async () => {
    try {
        await actionFn()
    } catch (e: any) {
        console.error('Error in ' + description, e)
        uiManager.showAlert(description, `Message: ${e.message}`, null, 'error')
    }
    uiManager.onClose()
}