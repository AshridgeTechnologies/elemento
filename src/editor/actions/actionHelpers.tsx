import React, {ChangeEvent, ReactNode, useState} from 'react'
import {Alert, AlertColor, AlertTitle} from '@mui/material'
import EditorManager from './EditorManager'
import {ASSET_DIR, projectFileName} from '../../shared/constants'

const OPFS_PROJECTS_DIR = 'Projects'

export type ShowAlertFn = (title: string, message: string, detail: React.ReactNode, severity: AlertColor) => void

export type Optional<T> = {
    [Property in keyof T]+?: T[Property]
}

export async function chooseDirectory(): Promise<FileSystemDirectoryHandle | null> {
    return window.showDirectoryPicker({id: 'elemento_editor', mode: 'readwrite'}).catch(e => {
            if (userCancelledFilePick(e)) {
                return null
            }
            throw e
        })
}

export async function validateDirectory(directory: FileSystemDirectoryHandle | null) {
    // @ts-ignore
    const dirEntry = await directory?.keys().next()
    return dirEntry.value ? 'Folder is not empty' : null
}

export async function validateDirectoryForOpen(directory: FileSystemDirectoryHandle | null) {
    if (!directory) return null
    try {
        await directory.getFileHandle(projectFileName)
        return null
    } catch (e) {
        return `Directory '${directory.name}' does not contain an Elemento project file`
    }
}

export const onChangeValue = (onChangeFn: (val: string) => void) => (event: ChangeEvent) => onChangeFn((event.target as HTMLInputElement).value)

export class UIManager {
    readonly onClose: () => void
    readonly showAlert: ShowAlertFn
    constructor(props: {onClose: () => void, showAlert: ShowAlertFn}) {
        this.onClose = props.onClose
        this.showAlert = props.showAlert
    }
}

type StateClass<T> = { new(props: { editorManager: EditorManager, uiManager: UIManager }, onUpdate: (t: T) => void): T }

export function useDialogState<T>(stateClass: StateClass<T>, editorManager: EditorManager, uiManager: UIManager) {
    const onUpdate = (newState: T) => setState(newState)
    const [state, setState] = useState(() => new stateClass({editorManager, uiManager}, onUpdate))
    return state
}

export const doAction = (uiManager: UIManager, description: string, actionFn: () => (void)) => async () => {
    try {
        await actionFn()
    } catch (e: any) {
        console.error('Error in ' + description, e)
        uiManager.showAlert(description, `Message: ${e.message}`, null, 'error')
    }
    uiManager.onClose()
}
export const userCancelledFilePick = (e: any) => /*e instanceof DOMException &&*/ e.name === 'AbortError'

export const internalProjectsDir = async () => {
    const opfsRoot = await navigator.storage.getDirectory()
    return opfsRoot.getDirectoryHandle(OPFS_PROJECTS_DIR, {create: true})
}
export const internalProject = async (name: string) => {
    const projectsDir = await internalProjectsDir()
    return projectsDir.getDirectoryHandle(name)
}
export const openFromGitHub = async (url: string, editorManager: EditorManager) => {
    const repoName = new URL(url).pathname.substring(1)
    const tidiedRepoName = repoName.replace(/\//, '_').replace(/\.git$/, '')
    const dateTime = new Date().toISOString().substring(0,19).replace(/[:\-T]/g, '')
    const dirName = `${tidiedRepoName}_${dateTime}`
    const projectsDir = await internalProjectsDir()
    const opfsDirectory = await projectsDir.getDirectoryHandle(dirName, {create: true})
    return editorManager.getFromGitHub(url, opfsDirectory)
}

export const createNewProject = async (name: string, editorManager: EditorManager) => {
    const projectsDir = await internalProjectsDir()
    const opfsDirectory = await projectsDir.getDirectoryHandle(name, {create: true})
    return editorManager.newProject(opfsDirectory)
}

export const validateProjectName = async (name: string | null) => {
    if (!name) return 'Name cannot be empty'
    if (name.includes('/')) return 'Name cannot contain /'
    const projectsDir = await internalProjectsDir()
    let existingProject: FileSystemDirectoryHandle
    try {
        await projectsDir.getDirectoryHandle(name)
        return 'A project with this name already exists'
    } catch (e) {
        return null
    }
}

export function AlertMessage({severity, removeAlert, title, message, detail}:
                                 { severity: AlertColor, removeAlert: VoidFunction, title: string, message: string, detail: ReactNode }) {
    return <Alert severity={severity} onClose={removeAlert}>
        <AlertTitle>{title}</AlertTitle>
        <p id="alertMessage">{message}</p>
        <p>{detail}</p>
    </Alert>
}

export const loadAppNames = async (projectDir: FileSystemDirectoryHandle): Promise<string[]> => {
    const distDir = await projectDir.getDirectoryHandle('dist')
    const clientDir = await distDir.getDirectoryHandle('client')
    const result: string[] = []
    // @ts-ignore
    for await (const [name] of clientDir.entries()) {
        if (name !== ASSET_DIR) result.push(name)
    }
    return result
}
