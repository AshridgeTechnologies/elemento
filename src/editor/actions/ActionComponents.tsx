import {
    Button,
    Dialog,
    DialogActions,
    DialogContent,
    DialogContentText,
    DialogTitle,
    IconButton,
    InputAdornment,
    Stack,
    TextField,
    TextFieldProps
} from '@mui/material'
import {Close} from '@mui/icons-material'
import React from 'react'
import {chooseDirectory} from './actionHelpers'
import {editorDialogContainer} from '../EditorElement'
import {mergeDeepRight} from 'ramda'
import {editorMenuPositionProps} from '../Editor'

export function CloseButton(props: { onClose: () => void }) {
    return <IconButton
        aria-label="close"
        onClick={props.onClose}
        sx={{
            position: 'absolute',
            right: 8,
            top: 8,
            color: (theme) => theme.palette.grey[500],
        }}
    >
        <Close/>
    </IconButton>
}

export function DialogTextField(props: TextFieldProps) {
    return <TextField
        margin="dense"
        fullWidth
        variant="standard"
        error={!!props.helperText}
        {...props}
    />
}

const dialogSlotProps = {
    root: {
        sx: {position: 'absolute'},
    },
    backdrop: {
        sx: {
            position: 'absolute',
            backgroundColor: 'rgba(0, 0, 0, 0.25)'
        }}

}

export function EditorActionDialog({title, content, fields, action, onCancel}:
                                 { title: string, content: React.ReactNode, fields: React.ReactNode, action: React.ReactNode, onCancel: VoidFunction }) {
    return (
        <Dialog onClose={onCancel} open={true} sx={{minWidth: 400}} container={editorDialogContainer()} slotProps={dialogSlotProps}>
            <DialogTitle>{title} <CloseButton onClose={onCancel}/></DialogTitle>
            <DialogContent>
                <DialogContentText>
                    {content}
                </DialogContentText>
                {fields}
            </DialogContent>
            <DialogActions>
                {action}
                <Button variant='outlined' onClick={onCancel}>Cancel</Button>
            </DialogActions>
        </Dialog>
    )

}

export function DirectoryInput(props: {
    id: string,
    label: string,
    value: FileSystemDirectoryHandle | null,
    onChange: (dir: FileSystemDirectoryHandle | null) => any,
    helperText: string | null
}) {

    const chooseDir = () => chooseDirectory().then(props.onChange)
    const inputProps = {
        readOnly: true,
        endAdornment:
            <InputAdornment position="end">
                <Button aria-label="choose folder" onClick={chooseDir}>Choose</Button>
            </InputAdornment>
    }
    return <Stack direction='row'>
        <TextField
            id={props.id} label={props.label}
            margin="dense"
            fullWidth
            variant="standard"
            value={props.value?.name ?? ''}
            error={!!props.helperText}
            helperText={props.helperText}
            InputProps={inputProps}
        />
    </Stack>
}
