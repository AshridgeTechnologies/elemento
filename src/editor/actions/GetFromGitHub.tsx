import React from 'react'
import {Button, InputAdornment, Stack, TextField} from '@mui/material'
import {DialogTextField, DirectoryInput, EditorActionDialog} from './ActionComponents'
import {
    doAction,
    onChangeValue,
    Optional,
    UIManager,
    useDialogState,
    userCancelledFilePick,
    validateDirectory
} from './actionHelpers'
import EditorManager from './EditorManager'

export class GetFromGitHubState {
    static initialState = () => ({
        url: '',
        directory: null,
        directoryError: null
    })

    constructor(private props: {editorManager: EditorManager, uiManager: UIManager},
                private onUpdate: (newState: GetFromGitHubState) => void,
                private state: {url: string, directory: FileSystemDirectoryHandle | null, directoryError: string | null} = GetFromGitHubState.initialState()) {
    }
    setUrl = (url: string) => this.updateWithChanges({url})
    setDirectory = (directory: FileSystemDirectoryHandle | null) => {
        return validateDirectory(directory)
            .then(directoryError => this.updateWithChanges({directory, directoryError}))
    }

    private withChanges(changes: Optional<typeof this.state>) {
        return new GetFromGitHubState(this.props, this.onUpdate, {...this.state, ...changes})
    }

    private updateWithChanges(changes: Optional<typeof this.state>) {
        this.onUpdate(this.withChanges(changes))
    }

    get url() { return this.state.url }
    get directory() { return this.state.directory }
    get directoryError() { return this.state.directoryError }
    get canDoAction() { return !!(this.url && this.directory && !this.directoryError) }

    onGet = () => this.props.editorManager.getFromGitHub(this.state.url, this.state.directory!)
}

export function GetFromGitHubDialog({editorManager, uiManager}: { editorManager: EditorManager, uiManager: UIManager }) {
    const state = useDialogState(GetFromGitHubState, editorManager, uiManager)
    const {url, directory, directoryError, canDoAction, onGet} = state

    const onChangeUrl = onChangeValue(state.setUrl)
    const onChangeDirectory = state.setDirectory

    const title = 'Get project from GitHub'
    const onGetAction = doAction(uiManager, title, onGet)
    return <EditorActionDialog
        onCancel={uiManager.onClose} title={title}
        content='Please enter the GitHub URL and a name for the new project.'
        fields={<>
            <DialogTextField
                autoFocus
                id='url'
                label='GitHub URL'
                value={url}
                onChange={onChangeUrl}/>

            <DirectoryInput
                id='projectDirectory'
                label='Folder for new project'
                value={directory}
                onChange={onChangeDirectory}
                helperText={directoryError}/>
        </>}
        action={<Button variant='outlined' onClick={onGetAction} disabled={!canDoAction}>Get</Button>}
    />
}
