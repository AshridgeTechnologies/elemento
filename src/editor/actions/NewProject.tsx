import React from 'react'
import {Button} from '@mui/material'
import {DirectoryInput, EditorActionDialog} from './ActionComponents'
import {doAction, Optional, UIManager, useDialogState, validateDirectory} from './actionHelpers'
import EditorManager from './EditorManager'

export class NewProjectState {
    static initialState = () => ({
        directory: null,
        directoryError: null
    })

    constructor(private props: {editorManager: EditorManager, uiManager: UIManager},
                private onUpdate: (newState: NewProjectState) => void,
                private state: {directory: FileSystemDirectoryHandle | null, directoryError: string | null} = NewProjectState.initialState()) {
    }
    setDirectory = (directory: FileSystemDirectoryHandle) => {
        return validateDirectory(directory)
            .then(directoryError => this.updateWithChanges({directory, directoryError}))
    }

    private withChanges(changes: Optional<typeof this.state>) {
        return new NewProjectState(this.props, this.onUpdate, {...this.state, ...changes})
    }

    private updateWithChanges(changes: Optional<typeof this.state>) {
        this.onUpdate(this.withChanges(changes))
    }

    get directory() { return this.state.directory }
    get directoryError() { return this.state.directoryError }
    get canDoAction() { return !!(this.directory && !this.directoryError) }

    onNew = () => this.props.editorManager.newProject(this.state.directory!)
}

export function NewProjectDialog({editorManager, uiManager}: { editorManager: EditorManager, uiManager: UIManager }) {
    const state = useDialogState(NewProjectState, editorManager, uiManager)
    const {directory, directoryError, canDoAction, onNew} = state

    const onChangeDirectory = state.setDirectory

    const title = 'Create New project'
    const onNewAction = doAction(uiManager, title, onNew)
    return <EditorActionDialog
        onCancel={uiManager.onClose} title={title}
        content='Please choose a directory to store the files for the new project.'
        fields={<>
            <DirectoryInput
                id='projectDirectory'
                label='Folder for new project'
                value={directory}
                onChange={onChangeDirectory}
                helperText={directoryError}/>
        </>}
        action={<Button variant='outlined' onClick={onNewAction} disabled={!canDoAction}>Create</Button>}
    />
}
