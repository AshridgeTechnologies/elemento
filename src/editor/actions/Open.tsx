import React from 'react'
import {Button, DialogContentText, List, ListItem, ListItemButton, ListItemText, Typography} from '@mui/material'
import {EditorActionDialog} from './ActionComponents'
import {
    chooseDirectory,
    doAction,
    internalProject, internalProjectsDir,
    Optional,
    UIManager,
    useDialogState, validateDirectoryForOpen
} from './actionHelpers'
import EditorManager from './EditorManager'
import {DiskProjectStore} from '../DiskProjectStore'



export class OpenState {
    static initialState = () => ({
        names: [] as string[] ,
        namesLoaded: false,
    })

    constructor(private props: {editorManager: EditorManager, uiManager: UIManager},
                private onUpdate: (newState: OpenState) => void,
                private state: {names: string[], namesLoaded: boolean} = OpenState.initialState()) {
        if (!this.state.namesLoaded) {
            internalProjectsDir().then( async (dir) => {
                const result: string[] = []
                // @ts-ignore
                for await (const [name] of dir.entries()) {
                    result.push(name)
                }
                return result
            }).then( this.setNames )
        }
    }

    setNames = (names: string[]) => this.updateWithChanges({names, namesLoaded: true})

    private withChanges(changes: Optional<typeof this.state>) {
        return new OpenState(this.props, this.onUpdate, {...this.state, ...changes})
    }

    private updateWithChanges(changes: Optional<typeof this.state>) {
        this.onUpdate(this.withChanges(changes))
    }

    get names() { return this.state.names }
    get canDoAction() { return true }

    onOpen = () => null
}

export function OpenDialog({editorManager, uiManager}: { editorManager: EditorManager, uiManager: UIManager }) {
    const state = useDialogState(OpenState, editorManager, uiManager)
    const {names, canDoAction, onOpen} = state

    const openProjectFromDisk = async () => {
        const dirHandle = await chooseDirectory()
        if (dirHandle) {
            const error = await validateDirectoryForOpen(dirHandle)
            if (error) {
                uiManager.showAlert('Open project', error, null, 'error')
            } else {
                await editorManager.openProject(dirHandle)
            }
        }
    }

    const title = 'Open project'
    const onOpenAction = doAction(uiManager, title, openProjectFromDisk)
    return <EditorActionDialog
        onCancel={uiManager.onClose} title={title}
        content='Choose a project stored in the browser'
        fields={<>
            <List dense sx={{maxHeight: 300, overflowY: 'auto', mt: 2, mb: 2}}>
                {names.map( (proj: string, ) => (
                    <ListItemButton key={proj} onClick={doAction(uiManager, title, async () => editorManager.openProject(await internalProject(proj)) )}>
                        <ListItemText primary={proj}/>
                    </ListItemButton>
                ))}
            </List>
            <DialogContentText sx={{borderTop: '1px solid lightgray', pt: 2}}>Or click Choose Folder to open a project on your computer disk</DialogContentText>
        </>}
        action={<Button variant='outlined' onClick={onOpenAction} disabled={!canDoAction}>Choose Folder</Button>}
    />
}
