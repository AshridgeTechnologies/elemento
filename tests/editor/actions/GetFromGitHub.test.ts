/**
 * @jest-environment jsdom
 */

import {GetFromGitHubDialog, GetFromGitHubState} from '../../../src/editor/actions/GetFromGitHub'
import EditorManager from '../../../src/editor/actions/EditorManager'
import {UIManager} from '../../../src/editor/actions/actionHelpers'
import {wait} from '../../testutil/testHelpers'
import {render, screen} from '@testing-library/react'
import '@testing-library/jest-dom'

import {createElement} from 'react'
import {actWait} from '../../testutil/rtlHelpers'
import userEvent from '@testing-library/user-event'

let container: any

let editorManager: EditorManager
let uiManager: UIManager
let onUpdate: (newState: GetFromGitHubState) => void
let state: GetFromGitHubState

beforeEach(() => {
    editorManager = {
        getProjectNames: jest.fn().mockResolvedValue(['Project A', 'Project B']),
        getFromGitHub: jest.fn().mockResolvedValue(undefined)
    } as unknown as EditorManager

    uiManager = new UIManager({onClose: jest.fn(), showAlert: jest.fn(), editorId: 'editor1'})

    onUpdate = jest.fn().mockImplementation( newState => state = newState)
    state = new GetFromGitHubState({editorManager}, onUpdate)
})


test('state sets project name', () => {
    state.setProjectName('Proj One')
    expect(state.directory).toBe('Proj One')
})

test('state sets url and leaves project name empty if not found', () => {
    state.setUrl('http://example.com')
    expect(state.url).toBe('http://example.com')
    expect(state.directory).toBe('')
})

test('state sets url and updates project name if empty or same as url', () => {
    state.setUrl('https://example.com/user/repo1')
    expect(state.url).toBe('https://example.com/user/repo1')
    expect(state.directory).toBe('repo1')
    state.setUrl('https://example.com/user/repo12')
    expect(state.url).toBe('https://example.com/user/repo12')
    expect(state.directory).toBe('repo12')
})

test('state sets url and leaves project name if different to url', () => {
    state.setProjectName('Proj One')
    state.setUrl('https://example.com/user/repo1')
    expect(state.url).toBe('https://example.com/user/repo1')
    expect(state.directory).toBe('Proj One')
})

test('validates project name not same as an existing name', async () => {
    await wait(10) // wait for xisting project names to load
    state.setProjectName('Proj One')
    expect(state.directoryError).toBeNull()
    state.setProjectName('Project B')
    expect(state.directoryError).toBe('There is already a project with this name')
})

test('can create when fields entered and no error', async () => {
    await wait(10) // wait for existing project names to load

    expect(state.canDoAction).toBe(false)
    state.setProjectName('Proj One')
    expect(state.canDoAction).toBe(false)
    state.setUrl('https://example.com/user/repo1')
    expect(state.canDoAction).toBe(true)
    state.setProjectName('Project B')
    expect(state.canDoAction).toBe(false)
})

test('calls onGet', async () => {
    state.setProjectName('Proj One')
    state.setUrl('https://example.com/user/repo1')
    state.onGet()
    expect(editorManager.getFromGitHub).toHaveBeenCalledWith('https://example.com/user/repo1', 'Proj One')
})

test('calls onGet and throws if error', async () => {
    editorManager = {
        getProjectNames: jest.fn().mockResolvedValue([]),
        getFromGitHub: jest.fn().mockRejectedValue(new Error('Cannot do this'))
    } as unknown as EditorManager
    state = new GetFromGitHubState({editorManager}, onUpdate)

    state.setProjectName('Proj One')
    state.setUrl('https://example.com/user/repo1')
    expect(state.onGet()).rejects.toThrow('Cannot do this')
})

test('dialog works with state', async () => {
    await actWait(() => ({container} = render(createElement(GetFromGitHubDialog, {editorManager, uiManager}))))
    expect(screen.getByRole('dialog')).toHaveTextContent('Get project from GitHub')

    const user = userEvent.setup()
    await user.type(screen.getByLabelText('GitHub URL'), 'https://example.com/user/Project A')
    expect(screen.getByLabelText('Project Name')).toHaveValue('Project A')
    expect(screen.getByRole('dialog')).toHaveTextContent('There is already a project with this name')

    expect(screen.getByRole('dialog').innerHTML).toMatchSnapshot()

    await user.type(screen.getByLabelText('Project Name'), '1')
    await user.click(screen.getByText('Get'))
    await wait(10)
    expect(editorManager.getFromGitHub).toHaveBeenCalledWith('https://example.com/user/Project A', 'Project A1')
    expect(uiManager.onClose).toHaveBeenCalled()
})

test('dialog shows alert if error in action', async () => {
    const originalConsoleError = console.error
    const error = new Error('Cannot do this')

    try {
        console.error = jest.fn()
        editorManager = {
            getProjectNames: jest.fn().mockResolvedValue([]),
            getFromGitHub: jest.fn().mockRejectedValue(error)
        } as unknown as EditorManager

        await actWait(() => ({container} = render(createElement(GetFromGitHubDialog, {editorManager, uiManager}))))
        const user = userEvent.setup()
        await user.type(screen.getByLabelText('GitHub URL'), 'https://example.com/user/Project One')
        await user.click(screen.getByText('Get'))
        expect(uiManager.showAlert).toHaveBeenCalledWith('Get project from GitHub', `Error: Cannot do this`, null, 'error')
        expect(uiManager.onClose).toHaveBeenCalled()
    } finally {
        expect(console.error).toHaveBeenCalledWith('Error in Get project from GitHub', error)
        console.error = originalConsoleError
    }
})
