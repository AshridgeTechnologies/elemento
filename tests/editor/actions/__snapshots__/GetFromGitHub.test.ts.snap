// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`dialog works with state 1`] = `"<h2 class="MuiTypography-root MuiTypography-h6 MuiDialogTitle-root css-rvoa5x-MuiTypography-root-MuiDialogTitle-root" id=":r0:">Get project from GitHub <button class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-e8z3iq-MuiButtonBase-root-MuiIconButton-root" tabindex="0" type="button" aria-label="close"><svg class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1umw9bq-MuiSvgIcon-root" focusable="false" aria-hidden="true" viewBox="0 0 24 24" data-testid="CloseIcon"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"></path></svg></button></h2><div class="MuiDialogContent-root css-kw13he-MuiDialogContent-root"><p class="MuiTypography-root MuiDialogContentText-root MuiTypography-body1 MuiDialogContentText-root css-ujudvl-MuiTypography-root-MuiDialogContentText-root">Please enter the GitHub URL and a name for the new project.</p><div class="MuiFormControl-root MuiFormControl-marginDense MuiFormControl-fullWidth MuiTextField-root css-xanwcj-MuiFormControl-root-MuiTextField-root"><label class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-standard MuiFormLabel-colorPrimary MuiFormLabel-filled Mui-focused MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-standard css-1ew92b2-MuiFormLabel-root-MuiInputLabel-root" data-shrink="true" for="url" id="url-label">GitHub URL</label><div class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-colorPrimary MuiInputBase-fullWidth Mui-focused MuiInputBase-formControl css-6h9cc6-MuiInputBase-root-MuiInput-root"><input aria-invalid="false" id="url" type="text" class="MuiInputBase-input MuiInput-input css-1yrc8ca-MuiInputBase-input-MuiInput-input" value="https://example.com/user/Project A"></div></div><div class="MuiStack-root css-m69qwo-MuiStack-root"><div class="MuiFormControl-root MuiFormControl-marginDense MuiFormControl-fullWidth MuiTextField-root css-xanwcj-MuiFormControl-root-MuiTextField-root"><label class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-standard MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-standard css-2hy7sk-MuiFormLabel-root-MuiInputLabel-root" data-shrink="false" for="projectDirectory" id="projectDirectory-label">Folder for new project</label><div class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-adornedEnd Mui-readOnly MuiInputBase-readOnly css-6h9cc6-MuiInputBase-root-MuiInput-root"><input aria-invalid="false" id="projectDirectory" readonly="" type="text" class="MuiInputBase-input MuiInput-input MuiInputBase-inputAdornedEnd Mui-readOnly MuiInputBase-readOnly css-1yrc8ca-MuiInputBase-input-MuiInput-input" value=""><div class="MuiInputAdornment-root MuiInputAdornment-positionEnd MuiInputAdornment-standard MuiInputAdornment-sizeMedium css-elo8k2-MuiInputAdornment-root"><button class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary css-1588512-MuiButtonBase-root-MuiButton-root" tabindex="0" type="button" aria-label="choose folder">Choose</button></div></div></div></div></div><div class="MuiDialogActions-root MuiDialogActions-spacing css-15fu35s-MuiDialogActions-root"><button class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-colorPrimary Mui-disabled MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-colorPrimary css-1btmbba-MuiButtonBase-root-MuiButton-root" tabindex="-1" type="button" disabled="">Get</button><button class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-colorPrimary css-1btmbba-MuiButtonBase-root-MuiButton-root" tabindex="0" type="button">Cancel</button></div>"`;
